pkg/api/mock/mock_api_implementation_test.go

```diff
@@ -0,0 +1,226 @@
+package mock_test
+
+import (
+	"context"
+	"testing"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+
+	clouddnsv1 "go.anx.io/go-anxcloud/pkg/apis/clouddns/v1"
+	corev1 "go.anx.io/go-anxcloud/pkg/apis/core/v1"
+	vlanv1 "go.anx.io/go-anxcloud/pkg/apis/vlan/v1"
+
+	"github.com/mitchellh/copystructure"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "go.anx.io/go-anxcloud/pkg/api/mock"
+)
+
+func TestMockAPI(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Mock API tests")
+}
+
+var _ = Describe("Mock API implementation", func() {
+
+	Context("CRUD Operations", Ordered, func() {
+		a := NewMockAPI()
+
+		testVLAN := &vlanv1.VLAN{Name: "abc", Locations: []corev1.Location{{Identifier: "test-location"}}}
+
+		dnsRecords := []types.Object{
+			&clouddnsv1.Record{Name: "a-record", Type: "A", RData: "1.1.1.1"},
+			&clouddnsv1.Record{Name: "txt-record", Type: "TXT", RData: "hello world"},
+		}
+
+		objects := []types.Object{
+			&clouddnsv1.Zone{Name: "mock.go-anxcloud.test"},
+			testVLAN,
+		}
+
+		objects = append(objects, dnsRecords...)
+
+		It("can Create resources", func() {
+			for _, o := range objects {
+				err := a.Create(context.TODO(), o)
+				Expect(err).ToNot(HaveOccurred())
+			}
+		})
+
+		It("supports not found errors on Get operations", func() {
+			err := a.Get(context.TODO(), &vlanv1.VLAN{Identifier: "nonexistent-identifier"})
+			Expect(err).To(MatchError(api.ErrNotFound))
+		})
+
+		It("can Get resources", func() {
+			vlan := &vlanv1.VLAN{Identifier: testVLAN.Identifier}
+			err := a.Get(context.TODO(), vlan)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(testVLAN).To(Equal(vlan))
+		})
+
+		It("can Update resources", func() {
+			copy, err := copystructure.Copy(testVLAN)
```

Comment:
I fear this is not a good test for this, this alternate one gives different results with real API vs mock API client:

```go
vlanUpdate := vlanv1.VLAN{
    Identifier: testVLAN.Identifier,
    Name:       "def",
}

err := a.Update(context.TODO(), &vlanUpdate)
Expect(err).ToNot(HaveOccurred())
Expect(vlanUpdate.Locations).To(Equal(testVLAN.Locations))
```

We mark a lot of fields with `omitempty`, resulting in them not being sent to the Engine, in turn resulting in those fields not being changed. This usage is an important one and mock client should behave as the real Engine API here

Reply:
Another detail in that example: Update requests most of the time get the full object in the response body, so `Locations` should be filled into `vlanUpdate` (as all the other attributes).